# Metabolic modelling to better understand the role of the microbiome in inflammatory bowel disease 

## Description of the project

Inflammatory bowel diseases (IBD), which include Crohn’s disease and ulcerative colitis, affect several million individuals worldwide. Crohn’s disease and ulcerative colitis are complex diseases that are heterogeneous at the clinical, immunological, molecular, genetic, and microbial levels. As part of the NIH (National Institutes of Health) integrative human microbiome project, a study was performed in which data at these various levels was collected in Crohn’s and ulcerative colitis disease patients. Host transcriptomics, metagenomics, proteomics and metabolomics data sets were generated, and they are publicly available (https://ibdmdb.org/).

In this project, approaches and tools such as context-specific modeling will be used to create a map of the metabolic activity that takes place in the gut microbiome only, identifying metabolites, enzymatic activity, gene expression, gene-protein-reaction rules, metabolic tasks, where you aim to include aspects relevant to the  IBD dataset. You intend to base this development on existing metabolic maps like the Recon3 set. To evaluate the usefulness of the resulting whole microbiome metabolic map you will use it for a data analysis pipeline that you will build to perform an integrative functional analysis to better understand the effect of the microbiome and derived metabolites in patients. For this pipeline, you will evaluate existing tools including your own and adapt where needed.

The evaluation dataset has information about the difference in metabolite and enzyme abundances in stool samples of patients and control subjects. In these stool samples, metagenomics and metabolomics are measured. Evaluating the microbiome genome full sequence data that are already analysed leads to a list of enzymes and abundance. You will try to use these data to compare patients and healthy people and functionally interpret the results. To do so you will try to map these results to full microbiome metabolic maps. The aim of this stage is to provide an easy-to-use interface that permits the mapping of their metabolites and/or genes (including KEGG orthologs or KOs) onto different types of molecular interaction networks. This network visualization can then be used to gain novel insights or assist people with the development of new hypotheses.

By creating the map/model of the gut microbiome, future analyses can use it to study how the different levels of the metabolites, enzymes can influence the development of different diseases in the gut, such as inflammatory bowel disease, Crohn’s disease and ulcerative colitis and at the same time how other parameters, including diet and drugs, affect the overall health of the gut.


## Research question

