---
title: "data preparation"
author: "ddedesener"
date: "02/07/21"
output:
 md_document:
    variant: markdown_github
always_allow_html: true
---
## Introduction 
In this workflow, differential gene expression analysis will be performed on host-transcriptomics data.

## R environment setup
First, we need to make sure all required packages 
```{r setup, warning=FALSE, message=FALSE}
# check if libraries are already installed > otherwise install it
if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager",repos = "http://cran.us.r-project.org")
if(!"rstudioapi" %in% installed.packages()) BiocManager::install("rstudioapi")
if(!"baySeq" %in% installed.packages()) BiocManager::install("baySeq")
if(!"DESeq2" %in% installed.packages()) BiocManager::install("DESeq2")
if(!"edgeR" %in% installed.packages()) BiocManager::install("edgeR")
if(!"bioDist" %in% installed.packages()) BiocManager::install("bioDist")
if(!"biomaRt" %in% installed.packages()) BiocManager::install("biomaRt")
if(!"dplyr" %in% installed.packages()) BiocManager::install("dplyr")
if(!"magrittr" %in% installed.packages()) BiocManager::install("magrittr")
if(!"EnhancedVolcano" %in% installed.packages()) BiocManager::install("EnhancedVolcano")
if(!"clusterProfiler" %in% installed.packages()) BiocManager::install("clusterProfiler")
if(!"KEGGREST" %in% installed.packages()) BiocManager::install("KEGGREST")
if(!"org.EcK12.eg.db" %in% installed.packages()) BiocManager::install("org.EcK12.eg.db")
if(!"org.Hs.eg.db" %in% installed.packages()) BiocManager::install("org.Hs.eg.db")

#load packages
library(rstudioapi)
library(baySeq)
library(DESeq2)
library(edgeR)
library(bioDist)
library(biomaRt)
library(dplyr)
library(ggplot2)
library(limma)
#library(gplots)
library(magrittr)
library(EnhancedVolcano)
library(org.Hs.eg.db)
#library(VennDiagram)
library(clusterProfiler)
library(KEGGREST)
library(org.EcK12.eg.db)
library(stringr)

# set your working environment to the location where your current source file is saved into.
setwd(dirname(rstudioapi::getSourceEditorContext()$path))
#include some functions adapted from ArrayAnalysis.org scripts
source("functions_ArrayAnalysis_v2.R")
WORK.DIR <- getwd()
```


## Data Preparations
The following section will prepar input data to be used in the analysis
```{r read data,warning=FALSE, message=FALSE}
#read  the input data file
mgxCount <- read.table(file = '../Data/ecs_relab_2', sep = '\t', header = TRUE)

#checking which samples have all zero values across all genes
#these sample should be removed otherwise there will be a problem when calculating estimate size factors
#idx <- which(colSums(mgxCount) == 0)
#CSMDRVXI MSM719ME  are samples who has all zero values so we remove them
#mgxCount <- mgxCount[ , -idx]

#read metadata file sample labels
sampleLabels <- read.table(file = "../Data/metadata", row.names=1,sep = '\t', stringsAsFactors = TRUE)
#apply same filtering to the metadata remove samples from sample labels metadata
#sampleLabels <- sampleLabels[-idx , ]

mgxCount <- subset(mgxCount, select=rownames(sampleLabels))

#add column names to the metadata file
colnames(sampleLabels) <- c( "sampleID", "biopsy_location", "disease")
#check whether sample names are in same order
all(colnames(mgxCount) == rownames(sampleLabels))

#select only biopsy_location and disease columns
#sampleLabels<-sampleLabels[, c(2,3)]
sampleLabels$disease <- relevel(sampleLabels$disease,ref="nonIBD")
#add an experimental group variable to sampleLabels
sampleLabels$group <- as.factor(sampleLabels$disease)
```

## Filtering Steps
We will apply some filtering process to filter out genes in the input data
```{r filtering,warning=FALSE, message=FALSE}
#remove genes which has all zero values for all samples then start DE analysis
mgxCount[is.na(mgxCount)] <- 0
nonzero <- rowSums(mgxCount) > 0
mgxCount %<>% .[nonzero,]

#############################CPM FILTERING#############################################
#aveLogCPM function compute average log2 counts-per-million for each row of counts.
#the below function is similar to log2(rowMeans(cpm(y, ...)))
#mean_log_cpm = aveLogCPM(mgxCount)
mean_log_cpm = log2(rowMeans(cpm(mgxCount)))

# We plot the distribution of average log2 CPM values to verify that our chosen presence threshold is appropriate. The distribution is expected to be bimodal, with a low-abundance peak representing non-expressed genes and a high-abundance peak representing expressed genes. The chosen threshold should separate the two peaks of the bimodal distribution. 
filter_threshold <- 0.75# we can try different threshold values
#jpeg(file="avgLogCpmDist.jpeg")#if you want to save the histogram uncomment the following command  
ggplot() + aes(x=mean_log_cpm) +
    geom_histogram(binwidth=0.2) +
    geom_vline(xintercept=filter_threshold) +
    ggtitle("Histogram of mean expression values")
#dev.off()#to save the plot to the file
#Having chosen our threshold, lets pick the subset of genes whose average expression passes that threshold.
keep_genes <- mean_log_cpm >= filter_threshold 
mgxCount <- mgxCount[keep_genes,]
#dim(mgxCount)#to check dimension of the data
###############################################################################
```
## T-tests
```{r}
# Separate CD, UC and nonIBD
cd <- sampleLabels[sampleLabels$disease=="CD",]
uc <- sampleLabels[sampleLabels$disease=="UC",]
nonIBD <- sampleLabels[sampleLabels$disease=="nonIBD",]

# select metagenomics data from CD, UC and nonIBD
metagenomics_cd <- subset(mgxCount, select= rownames(cd))
metagenomics_uc <- subset(mgxCount, select= rownames(uc))
metagenomics_nonIBD <- subset(mgxCount, select= rownames(nonIBD))
```

```{r}
# Compute statistical significance (using t-test)
pvalue_cd = NULL # Empty list for the p-values
padjust_cd = NULL
tstat_cd = NULL # Empty list of the t test statistics
regulation_cd = NULL

for(i in 1 : nrow(metagenomics_nonIBD)) { # For each gene : 
	x = metagenomics_nonIBD[i,] # control of gene number i
	y = metagenomics_cd[i,] # CD of gene number i
	
	# Compute t-test between the two conditions
	t = t.test(x, y)
	
	# Put the current p-value in the pvalues list
	pvalue_cd[i] = t$p.value
	# Put the current t-statistic in the tstats list
	tstat_cd[i] = t$statistic
	
	if (rowMeans(x)<rowMeans(y))
	{
	  regulation_cd[i] = "Up-regulated"
	}
	else
	{
	  regulation_cd[i] = "Down-regulated"
	}
}

padjust_cd = p.adjust(pvalue_cd, method="BH", n= nrow(metagenomics_nonIBD))
```


```{r}
# Compute statistical significance (using t-test)
pvalue_uc = NULL # Empty list for the p-values
padjust_uc = NULL
tstat_uc = NULL # Empty list of the t test statistics
regulation_uc = NULL

for(i in 1 : nrow(metagenomics_nonIBD)) { # For each gene : 
	x = metagenomics_nonIBD[i,] # control of gene number i
	y = metagenomics_uc[i,] # CD of gene number i
	
	# Compute t-test between the two conditions
	t = t.test(x, y)
	
	# Put the current p-value in the pvalues list
	pvalue_uc[i] = t$p.value
	# Put the current t-statistic in the tstats list
	tstat_uc[i] = t$statistic
	
	if (rowMeans(x)<rowMeans(y))
	{
	  regulation_uc[i] = "Up-regulated"
	}
	else
	{
	  regulation_uc[i] = "Down-regulated"
	}
}

padjust_uc = p.adjust(pvalue_uc, method="BH", n= nrow(metagenomics_nonIBD))

```

## KEGG pathway over-representation analysis
Codes for KEGG organisms can be found here (https://www.genome.jp/kegg/catalog/org_list.html)
```{r}
deg.CD <- metagenomics_cd[which(pvalue_cd<0.05),]
regulation.deg.CD <- regulation_cd[which(pvalue_cd<0.05)]
deg.UC <- metagenomics_uc[which(pvalue_uc<0.05),]
regulation.deg.UC <- regulation_uc[which(pvalue_uc<0.05)]
```

```{r}
# Subset Bacteroides xylanisolvens
deg.CD.xyl <- deg.CD[grepl("xylanisolvens", rownames(deg.CD), fixed = TRUE),]
deg.UC.xyl <- deg.UC[grepl("xylanisolvens", rownames(deg.UC), fixed = TRUE),]
```

```{r}
deg.CD.xyl$Gene.Family <- rownames(deg.CD.xyl)
deg.UC.xyl$Gene.Family <- rownames(deg.UC.xyl)
```

```{r}
# Split
deg.CD.xyl[c('Gene.Family', 'Name.Organism')] <- str_split_fixed(deg.CD.xyl$Gene.Family, ': ', 2)
deg.UC.xyl[c('Gene.Family', 'Name.Organism')] <- str_split_fixed(deg.UC.xyl$Gene.Family, ': ', 2)
```


```{r}
# Convert EC numbers to Entrez IDs
gene <- clusterProfiler::bitr(deg.CD.xyl$Gene.Family, fromType = "ENZYME", toType = "ENTREZID", OrgDb = org.EcK12.eg.db)

# Convert Entrez IDs to KEGG IDs
geneList <- sub("^", "ncbi-geneid:", gene[,2])
geneList <- keggConv("eco", geneList)
```

```{r}
# Remove preceding 'eco:'
geneList <- gsub("eco:", "", geneList)

kk <- enrichKEGG(gene         = geneList,
                 organism     = 'eco',
                 pvalueCutoff = 0.05)
head(kk)
```
